kafka-streams/src/main/java/Account.java

```diff
@@ -0,0 +1,152 @@
+import com.google.gson.JsonObject;
+
+import java.util.Objects;
+
+public class Account {
+
+    private final String giroNumber;
+    private double balance;
+    private final String currency;
+    private long valid_from;
+    private long valid_to;
+
+    private double averageOfExpenses; //- negative change in balance
+    private int numberOfExpenses;
+
+    private Account(String giroNumber, String currency, double balance, String valid_from, String valid_to) {
+        this.giroNumber = giroNumber;
+        this.currency = currency;
+        this.balance = balance;
+        this.valid_from = CustomDateFormatter.getTimeStampFromDateInSeconds(valid_from);
+        this.valid_to = CustomDateFormatter.getTimeStampFromDateInSeconds(valid_to);
+
+        averageOfExpenses = 0;
+        numberOfExpenses = 0;
+    }
+
+    private void refreshAverageOfExpenses(double newExpenseValue) {
+        averageOfExpenses += newExpenseValue;
+        numberOfExpenses += 1;
+        averageOfExpenses = averageOfExpenses / numberOfExpenses;
+    }
+
+    public void refresh(Account newAccount) {
+        if (giroNumber.equals(newAccount.giroNumber)) {
+            balance = newAccount.balance;
+            valid_from = newAccount.valid_from;
+            valid_to = newAccount.valid_to;
+            refreshAverageOfExpenses(newAccount.balance - balance);
+        }
+    }
+
+    static Account createNewAccountFromJson(JsonObject json) {
+        String giroNumber = json.get("GIRONUMBER").getAsString();
+        double amount = Double.parseDouble(json.get("AMOUNT").getAsString());
+        String currency = json.get("CURRENCY").getAsString();
+        String valid_from_date = json.get("VALID_FROM").getAsString();
+        String valid_to_date = json.get("VALID_TO").getAsString();
+        return new Account(giroNumber,currency,amount, valid_from_date, valid_to_date);
+    }
+
+    public boolean searchPatternInExpenses(Account newAccount) {
+        boolean anomaly = false;
+        double difference = newAccount.balance - balance;
+        if (difference < 0) { // only for withdraws
+            double deviation = Math.pow(difference - Math.abs(averageOfExpenses),2);
+            if (deviation > Math.abs(averageOfExpenses) + 0.5 * getAnomalyLimit(currency)) anomaly = true; // Predefined threshold
```

Comment:
Can you elaborate on this bit? How does it work?

Reply:
Actually we can still work on this, but for now the idea was to try to catch those withdraws from the account which are outliers. E.g too different from the mean of the previous spendings.
Bc we have only one new value every time, we need to compute the mean iteratively. From this we can calculate std dev.
Here we can try to improve by some kind of moving threshold, which can take the number of transactions into account, bc now it is just a hardcoded one, which may be not too effective in every cases. N=~10 vs N= 1000+